Dev/bap wps onboarding 3

tomereli opened the pull request:
Add support bAPs for allowing other agents to onboard the prplMesh agent using wireless backhaul:
- Use Intel hostapd mesh mode instead of multi_ap for VAPs configuration
- Add WPS PBC implementation using hostapd upstream support
- Remove backhaul vaps unused code

PR in prplwrt iwlwav - https://git.prpl.dev/prplmesh/iwlwav/-/merge_requests/12

Fixes #1070 
Fixes #347

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -87,22 +87,23 @@ static void copy_vaps_info(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
         if (radio_vaps.find(vap_id) != radio_vaps.end()) {
-            const auto &curr_vap = radio_vaps.at(vap_id);
+            const auto &curr_vap  = radio_vaps.at(vap_id);
+            sMacAddr curr_vap_mac = network_utils::mac_from_string(curr_vap.mac);
```
This change is unrelated to the rest, and it has limited merit (it's not necessary for this commit). If you want to keep it, it should be a separate commit.

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -258,20 +259,6 @@ void ap_manager_thread::after_select(bool timeout)
-            backhaul_vaps_list.remove_if([&](const bwl::backhaul_vap_list_element_t &element) {
-                return (element.type != current_type);
-            });
-
```
This change is not explained in the commit message, and should *definitely* be a separate commit.

tomereli replied:
obsolete since we removed the backhaul vaps list

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
     return true;
 }
 
+static bool set_vap_multiap_mode(std::vector<std::string> &vap_hostapd_config,
```
This should have been done in 3 separate commits:
1. Refactor set_vap_multiap_mode into a separate function without changing 
2. Add mesh_mode
3. Add set_multiap_wps

However, that's probably difficult to do after the fact, and it's mostly useful for reviewing, not for longer-term history. So OK.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+        LOG(ERROR) << "Failed to get VAP ifname (bss)";
+        return false;
+    }
+    backhaul_vaps_list.remove_if([&](const bwl::backhaul_vap_list_element_t &element) {
```
The same should also be done for the VAPs that are disabled later. So maybe it would be better to just clear backhaul_vaps_list completely and add to them again?

tomereli replied:
removed backhaul vaps list completely.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    });
+    bool backhaul  = !!(bss_type & WSC::eWscVendorExtSubelementBssType::BACKHAUL_BSS);
+    bool fronthaul = !!(bss_type & WSC::eWscVendorExtSubelementBssType::FRONTHAUL_BSS);
+    bool teardown  = (bss_type == WSC::eWscVendorExtSubelementBssType::TEARDOWN);
```
If this is true, it's an error. Teardown is handled differently in the bwl interface (by giving an empty bssinfo list).

Maybe it would be better to split up the fronthaul and backhaul as separate booleans in the bssinfo structure, then we don't create confusion about the teardown and bSTA bits, and we don't have to "parse" the bss_type in every bwl implementation. But that idea got shot down when I originally proposed it, so I guess it will still get shot down now :-)

tomereli replied:
Done, but as a new commit.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    hostapd_config_set_value(vap_hostapd_config, "mesh_mode", backhaul ? "bAP" : "fAP");
+    hostapd_config_set_value(vap_hostapd_config, "sFourAddrMode", backhaul ? "1" : "");
+    hostapd_config_set_value(vap_hostapd_config, "max_num_sta", backhaul ? "1" : "");
+    hostapd_config_set_value(vap_hostapd_config, "start_disabled", teardown ? "1" : "0");
```
So this shouldn't be done - the teardown elements are just missing from bssinfos.

tomereli replied:
Removed

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    // https://github.com/prplfoundation/prplMesh/issues/889)
+    if (fronthaul && backhaul) {
+        LOG(ERROR) << "Not supporting hybrid VAP";
+        return false;
```
Since the return value is ignored by the caller, we should at least still set the config to something. Probably disable the VAP is best.

Alternatively, we should handle the return value in the caller.

tomereli replied:
Correct, fixed to check the return value in the caller.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    std::map<std::string, std::vector<std::string>> &hostapd_config_vaps)
+{
+    // Find first fAP VAP
+    auto fap = std::find_if(hostapd_config_vaps.begin(), hostapd_config_vaps.end(),
```
I think it's easier to understand if this search is done while we're anyway iterating over bss_conf_list in update_vap_credentials(). Add
```
   auto first_fronthaul_vap = hostapd_config_vaps.end();
```
and set it in the loop.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,6 +934,96 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    }
+
+    // Find first bAP VAP
+    auto bap = std::find_if(hostapd_config_vaps.begin(), hostapd_config_vaps.end(),
```
Can't we use backhaul_vaps_list for this?

If not, same comment as above.

tomereli replied:
Completely different now, no backhaul vaps list.

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -164,16 +134,6 @@ void ap_manager_thread::ap_manager_config(ap_manager_conf_t &conf)
-        bwl::backhaul_vap_list_element_t element = {};
-        element.bssid = network_utils::mac_to_string(conf.backhaul_vaps_bssid[i]);
-        element.type  = bwl::eBackhaulVapType(i % 3);
-        backhaul_vaps_list.push_back(std::move(element));
```
I wonder if backhaul_vaps_list couldn't be removed entirely if we instead replace VAPElement with sBssInfoConf everywhere. Then the front/backhaul info is in there directly.

But no need to do that in this PR.

tomereli replied:
Removed entirely

arnout commented on common/beerocks/bwl/include/bwl/ap_wlan_hal.h:
```diff
@@ -164,12 +164,14 @@ class ap_wlan_hal : public virtual base_wlan_hal {
     virtual bool
-    update_vap_credentials(std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list) = 0;
+    update_vap_credentials(std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list,
+                           std::list<bwl::backhaul_vap_list_element_t> &backhaul_vaps_list) = 0;
```
I don't like this. The BWL should be an abstraction of the interface to wireless. It's not right that we rely on a bwl function to update an internal structure of ap_manager_thread. So unless the bwl actually does something with that backhaul vaps list that the ap_manager_thread can't do itself, I don't think it's right.

tomereli replied:
Done

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,8 +934,99 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    if (!hostapd_config_get_value(bap->second, "wpa_passphrase", backhaul_passphrase)) {
+        LOG(ERROR) << "Failed to get " << bap->first << "wpa_passphrase";
+        return false;
+    }
```
If we get the appropriate sBssInfo structures while iterating from them, we don't need this hack of parsing the config again, and this function can never fail. Would be much better.

tomereli replied:
Done

arnout commented on common/beerocks/bcl/source/beerocks_ucc_listener.cpp:
```diff
@@ -562,6 +564,22 @@ void beerocks_ucc_listener::handle_wfa_ca_command(const std::string &command)
         break;
     }
+    case eWfaCaCommand::START_WPS_REGISTRATION: {
+        std::unordered_map<std::string, std::string> params{};
```
I believe the band and wpsconfigmethod arguments are mandatory, so we can just as well add them here.

tomereli replied:
Correct. Changed.

arnout commented on common/beerocks/bcl/source/beerocks_ucc_listener.cpp:
```diff
@@ -562,6 +564,22 @@ void beerocks_ucc_listener::handle_wfa_ca_command(const std::string &command)
+            reply_ucc(eWfaCaStatus::INVALID, err_string);
+            break;
+        }
+        if (!handle_start_wps_registration(params, err_string)) {
```
... and pass them explicitly as arguments here instead of the generic "params".

Note that I would also do the string->enum conversion here.

(nitpick)

tomereli replied:
Not passing params anymore, I'm passing the band

arnout commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -389,6 +389,7 @@ enum eFreqType {
     FREQ_5G     = 1,
     FREQ_58G    = 2,
     FREQ_24G_5G = 3,
+    FREQ_60G    = 4,
```
This PR doesn't end up using it, so why add it?

tomereli replied:
removed

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml:
```diff
@@ -201,6 +201,10 @@ cACTION_APMANAGER_WIFI_CREDENTIALS_UPDATE_REQUEST:
 
+cACTION_APMANAGER_START_WPS_PBC_REQUEST:
+  _type: class
+  bssid: sMacAddr
```
I believe we just agreed to use interface names everywhere, so not ifindex, mac address, radio_uid or vap_id. In my view, "everywhere" includes the internal messages.

In other words, I think we should use hostap_iface here instead.

tomereli replied:
removed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -47,6 +47,22 @@ class backhaul_manager : public btl::transport_socket_thread {
+     * @param freq radio frequency to search
+     * @return radio mac of the found slave if found, otherwise empty string
+     */
+    const std::string freq_to_radio_mac(enum eFreqType freq) const;
```
I really don't like this approach that is taken in master::db of using a MAC address as an identifier. We're doing C++ here, not tcl, so not everything is a string - we have actual *objects* to manipulate. In other words, you can just return a `std::shared_ptr<sRadioInfo>`.

tomereli replied:
I think its easier to use the radio_mac instead of the sRadioInfo.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -47,6 +47,22 @@ class backhaul_manager : public btl::transport_socket_thread {
+     * @param bssid bssid on which to start WPS
+     * @return true on success, false on failure
+     */
+    bool start_wps_pbc(const sMacAddr &radio_mac, const sMacAddr &bssid);
```
(nitpick) it makes more sense to me to put the logic of which VAP to start WPS on *inside* this function.

And in keeping with my previous comment, the radio should be passed as a `const sRadioInfo&`.

tomereli replied:
Refactored.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -3592,4 +3592,45 @@ bool backhaul_manager::add_ap_he_capabilities(const sRadioInfo &radio_info)
     return true;
 }
 
+const std::string backhaul_manager::freq_to_radio_mac(enum eFreqType freq) const
```
(nitpick) although the type is called eFreqType, I think it makes more sense to call the variable `band` instead of `freq`.

Also, AFAIK, our coding style prescribes to drop the `enum`.

tomereli replied:
You are correct, fixed.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -186,8 +199,50 @@ bool agent_ucc_listener::handle_start_wps_registration(
+    auto freq      = band_to_freq(band->second);
+    auto radio_mac = m_backhaul_manager_ctx.freq_to_radio_mac(freq);
+    if (radio_mac.empty()) {
+        err_string = "Failed to get radio mac for " + band->second;
```
(nitpick) the CAPI errors should be relevant to the outside world. In this case:
```suggestion
        err_string = "Failed to get radio for " + band->second;
```

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -186,8 +199,50 @@ bool agent_ucc_listener::handle_start_wps_registration(
+
+    auto it = vaps_map.find(radio_mac);
+    if (it == vaps_map.end()) {
+        err_string = "radio_mac " + radio_mac + " not found";
```
... and here the same message, but a LOG(ERROR) with the internal message.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -186,8 +199,50 @@ bool agent_ucc_listener::handle_start_wps_registration(
+    if (backhaul_bssid == network_utils::ZERO_MAC) {
+        err_string = "No bAP found for radio_mac " + radio_mac;
+        return false;
+    }
```
We shouldn't do this. If PBC is triggered, it is *simultaneously* for fronthaul and backhaul. Therefore, we should trigger it even if no backhaul AP is configured.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -186,8 +199,50 @@ bool agent_ucc_listener::handle_start_wps_registration(
+    }
+    if (fronthaul_bssid == network_utils::ZERO_MAC) {
+        err_string = "No fAP found for radio_mac " + radio_mac;
+        return false;
```
This is a tricky one... It is possible that you have fronthaul only configured on one radio and backhaul only on another. Therefore, it is certainly not an error if no fronthaul is configured on this radio. Clearly we should not start WPS on this radio in that case, so we still have to return, but without error.

That said, it *is* an error if no fronthaul is configured on *any* radio. So to be completely correct, we should loop over all radios (of that band) and trigger an error if there is no fronthaul BSS on any of them.

But that is going a bit far, so let's just say that we don't consider it an error if no fronthaul is configured.

arnout reviewed (changes requested):
There is an inconsistency now: the logic to select the VAP for WPS is duplicated. When configuring it, the logic is inside the bwl implementation, but when WPS is triggered, the logic is in backhaul_manager (well, actually in agent_ucc_listener at the moment but I don't think that that's right). It would make more sense if the logic was completely in backhaul_manager. But that is not possible at the moment, because the configuration is handled by the son_slave while the WPS handling is done by backhaul_manager... So, it would actually make more sense if all the logic in either son_slave or ap_manager (depends on whether ap_manager is supposed to be nothing more than a bridge to bwl, or if it really should contain logic - I believe it should be just a bridge).

In a similar vein, the backhaul credentials should really be set on *all* radios which have WPS enabled, not only on the radios that have fronthaul (and therefore WPS) configured. So that would mean that when backhaul credentials are set, they have to be distributed to all slaves. And it probably means we need a separate bwl function for setting backhaul credentials for wps. That, however, is way out of scope for this PR. So let's create an issue for it.

Since I'll want to think a bit more about these two main issues, I'm marking it as changes requested.

The rest of my comments are smaller coding style issues.

arnout commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -55,9 +55,17 @@ struct WiFiChannel {
     bool radar_affected;
 };
 
+enum eVapType : uint8_t {
```
(nitpick) why not enum class?

tomereli replied:
removed altogether, added boolean values to VAPElement.

arnout commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -55,9 +55,17 @@ struct WiFiChannel {
+    VAP_TYPE_FRONTHAUL = 0,
+    VAP_TYPE_BACKHAUL  = 1,
+    VAP_TYPE_HYBRID    = 3,
+    VAP_TYPE_INVALID   = 4,
```
(nitpick) R2 will add more VAP types (R2-only and R1-only backhaul). So maybe we should use a bigger number already.

Why do we have values here anyway?

arnout replied:
Actually, since we now have bssinfo splitting fronthaul and backhaul as booleans, it would make sense to do the same in the VapElement.

tomereli replied:
Done - removed the enum.

arnout commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -825,6 +825,49 @@ bool base_wlan_hal_dwpal::refresh_radio_info()
+    }
+
+    std::string mesh_mode = std::string(mode);
+    if (mesh_mode.find("bAP") != std::string::npos) {
```
Shouldn't this be *exactly* bAP? I.e., `mesh_mode=foobAPbar` should not match, right?

In other words:
```suggestion
    if (mesh_mode == "bAP") {
```

Note: I know that the rest of this file uses this find pattern, but I don't think that's a reason to continue doing this weird and incorrect thing.

tomereli replied:
No, actually the value returned from hostapd is `mesh_mode=bAP (1)`, this is why I used `std::find`.

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -90,58 +88,17 @@ copy_vaps_info(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
             vaps[i].mac = network_utils::mac_from_string(curr_vap.mac);
             mapf::utils::copy_string(vaps[i].ssid, curr_vap.ssid.c_str(),
                                      beerocks::message::WIFI_SSID_MAX_LENGTH);
-            vaps[i].backhaul_vap = bVAP;
```
Shouldn't there be something here like
```
     vaps[i].backhaul_vap = (curr_vap.type == BACKHAUL || curr_vap.type == HYBRID);
```
?

tomereli replied:
Obsolete

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -90,58 +88,17 @@ copy_vaps_info(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
     }
 }
 
-static bool enable_backhaul_vap(
```
Your original commit message had a very good explanation of why this could be removed. Please keep that explanation.

tomereli replied:
Done. Thanks :)

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -1681,7 +1577,7 @@ bool ap_manager_thread::handle_ap_enabled(int vap_id)
     mapf::utils::copy_string(notification->vap_info().ssid, vap_info.ssid.c_str(),
                              beerocks::message::WIFI_SSID_MAX_LENGTH);
-    notification->vap_info().backhaul_vap = bVAP;
+    notification->vap_info().backhaul_vap = (vap_info.type == bwl::eVapType::VAP_TYPE_BACKHAUL);
```
```suggestion
    notification->vap_info().backhaul_vap = (vap_info.type == bwl::eVapType::VAP_TYPE_BACKHAUL || vap_info.type == bwl::eVapType::VAP_TYPE_HYBRID);
```

tomereli replied:
Obsolete since we use backhaul and fronthaul boolean values.

arnout commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -55,9 +55,17 @@ struct WiFiChannel {
 };
 
+enum eVapType : uint8_t {
+    VAP_TYPE_FRONTHAUL = 0,
```
This will always be fully scoped, so the VAP_TYPE_ part isn't needed.

Also, doesn't our coding style say this should be lower_snake_case? This file actually uses a mix of UPPER_SNAKE_CASE and UpperCamelCase, so I don't know what is the best now...

tomereli replied:
Deleted.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -95,10 +95,6 @@ sPlatformSettings:
   backhaul_max_vaps: uint8_t 
   backhaul_network_enabled: uint8_t 
   backhaul_preferred_radio_band: uint8_t 
-  backhaul_vaps_bssid:
```
Please explain in the commit message why this can be removed as well.

tomereli replied:
I added the original explanation, all the backhaul vaps are not used anymore. We have the backhaul and fronthaul booleans in the VAPElement.

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -863,7 +868,8 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
         }
 
-        ap_wlan_hal->update_vap_credentials(bss_info_conf_list);
+        ap_wlan_hal->update_vap_credentials(bss_info_conf_list, backhaul_wps_ssid,
+                                            backhaul_wps_passphrase);
```
(nitpick) I would have done this change (adding backhaul credentials to update_vap_credentials()) in a separate commit from the mesh_mode change, but you've already done so much rebasing that I don't want to ask you to split this off too :-)

tomereli replied:
You know that when you ask nicely I can't say no, so done :)

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,8 +934,53 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+                                 const std::string &backhaul_wps_ssid,
+                                 const std::string &backhaul_wps_passphrase)
+{
+    std::string bssid, ifname;
+    if (!hostapd_config_get_value(vap_hostapd_config, "bssid", bssid)) {
+        LOG(ERROR) << "Failed to get VAP BSSID";
+        return false;
+    }
+    if (!hostapd_config_get_value(vap_hostapd_config, "bss", ifname)) {
+        LOG(ERROR) << "Failed to get VAP ifname (bss)";
+        return false;
+    }
```
This is only used for a debug message, I don't think it's worth keeping it just for that.

tomereli replied:
Without the bssid or the interface name there is no much use of the debug message at all, so I prefer to keep them.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,8 +934,53 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    hostapd_config_set_value(vap_hostapd_config, "max_num_sta", backhaul ? "1" : "");
+    hostapd_config_set_value(vap_hostapd_config, "start_disabled", teardown ? "1" : "0");
+    if (fronthaul) {
+        // Oddly enough, multi_ap_backhaul_wpa_passphrase has to be quoted, while wpa_passphrase does not...
```
I wonder, shouldn't we also set wps_mode=2 in the fronthaul case and wps_mode=0 in the backhaul case? I.e., make sure WPS is only enabled on fronthaul.

I know that for upstream hostapd this wlll be needed, but I'm not sure for Intel hostapd.

tomereli replied:
You are absolutely right, I've been meaning to fix this and forgot - thank you for paying attention!
Its called `wps_state` though.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -169,6 +169,41 @@ bool agent_ucc_listener::send_cmdu_to_destination(ieee1905_1::CmduMessageTx &cmd
+        return false;
+    }
+
+    auto it = vaps_map.find(radio_mac);
```
Ah, this is why radio_mac is easier than sRadioInfo... It's the vaps_map that is broken :-)

tomereli replied:
Perhaps we can fix it then in a different PR :)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -3606,4 +3606,42 @@ bool backhaul_manager::add_ap_he_capabilities(const sRadioInfo &radio_info)
+        return false;
+    }
+
+    auto it = std::find_if(
+        slaves_sockets.begin(), slaves_sockets.end(),
+        [&](std::shared_ptr<sRadioInfo> slave) { return slave->radio_mac == radio_mac; });
+    if (it == slaves_sockets.end()) {
+        LOG(ERROR) << "couldn't find slave for radio mac " << radio_mac;
+        return false;
+    }
```
Just to be clear: this is why I think using the radio_mac as the identifier is wrong: it requires looking up the sRadioInfo again, while we already did that. Also, if we ever want to start pbc on all radios, then we will not be able to reuse this function.

That said, I'm OK with merging it as is now. I guess the private struct didn't work out after all.

tomereli replied:
We we should do is perhaps make the agent_ucc_listener friend class of the backhaul manager, then we can just use it as is even though it is private. What do you think (anyway its for a different PR) ?

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -875,7 +875,13 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+        if (!request) {
+            LOG(ERROR) << "addClass cACTION_APMANAGER_START_WPS_PBC_REQUEST failed";
+            return false;
+        }
```
It's a bit silly that we actually include this empty class, but OK.

tomereli replied:
Its leftovers from back when we passed the bssid in the message. It can be removed now.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1675,6 +1675,34 @@ bool ap_wlan_hal_dwpal::generate_connected_clients_events()
 
+bool ap_wlan_hal_dwpal::start_wps_pbc()
+{
+    refresh_vaps_info(beerocks::IFACE_RADIO_ID);
```
Why is it needed to refresh vaps info? Shouldn't it be up to date at this point?

Please add a comment to explain this.

tomereli replied:
I'm actually not sure if its up to date, I don't know when we start_wps_pbc() will be called so its better to be on the safe side. It doesn't happen too often.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1675,6 +1675,34 @@ bool ap_wlan_hal_dwpal::generate_connected_clients_events()
+                                   vap.second.type != eVapType::VAP_TYPE_FRONTHAUL) {
+                                   return false;
+                               }
+                               return true;
```
```suggestion
                               return (!vap.second.ssid.empty() &&
                                   vap.second.type == eVapType::VAP_TYPE_FRONTHAUL);
```

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1675,6 +1675,34 @@ bool ap_wlan_hal_dwpal::generate_connected_clients_events()
+bool ap_wlan_hal_dwpal::start_wps_pbc()
+{
+    refresh_vaps_info(beerocks::IFACE_RADIO_ID);
+    // start wps on the first fronthaul vap
```
I think it's worth to explain in the commit message that specifically dwpal doesn't support WPS on multiple interfaces, so we take the first fronthaul interface.

tomereli replied:
Done, updated the commit message + comment.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -501,6 +501,12 @@ bool ap_wlan_hal_nl80211::generate_connected_clients_events()
+bool ap_wlan_hal_nl80211::start_wps_pbc()
+{
+    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED!";
+    return true;
```
```suggestion
    LOG(DEBUG) << "Start WPS PBC on interface " << m_radio_info.iface_name;
    std::string cmd = "WPS_PBC";
    if (!wpa_ctrl_send_msg(cmd)) {
        LOG(ERROR) << "start_wps_pbc() failed!";
        return false;
    }
    return true;
```

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -659,12 +659,23 @@ bool master_thread::autoconfig_wsc_add_m2(WSC::m1 &m1,
+            cfg.bss_type = WSC::eWscVendorExtSubelementBssType::BACKHAUL_BSS;
+        } else {
+            cfg.bss_type = WSC::eWscVendorExtSubelementBssType::TEARDOWN;
+        }
```
NACK this!
- We *never* should set TEARDOWN here, teardown is handled in the else branch below.
- We should simply set the hybrid mode if hybrid mode is requested. If we don't do this, controller certification *will* fail.

```suggestion
        cfg.bss_type = 0;
        if (bss_info_conf->fronthaul) {
            cfg.bss_type |= WSC::eWscVendorExtSubelementBssType::FRONTHAUL_BSS;
        }
        if (bss_info_conf->backhaul) {
            cfg.bss_type |= WSC::eWscVendorExtSubelementBssType::BACKHAUL_BSS;
        }
```

tomereli replied:
You're right, fixed.

arnout commented on ci/certification/netgear-rax40.yml:
```diff
@@ -98,6 +98,26 @@ MAP-4.4.1_ETH:netgear-rax40:
     - when: manual
       allow_failure: true
 
+MAP-4.2.3_ETH_FH5GL:netgear-rax40:
```
These are already there, we just merged a commit that adds all certification tests as jobs. So this commit should be dropped.

arnout commented on tests/capi.py:
```diff
@@ -159,6 +159,18 @@ def cmd_reply(self, command: str, verbose: bool = False) -> Dict[str, str]:
             self.send_cmd(command)
             return self.get_reply(verbose)
 
+    def start_wps_registration(self, band: str):
```
```suggestion
    def start_wps_registration(self, band: str) -> None:
```

arnout reviewed (approved):
One big NACK, one thing which will make gitlab CI fail, and a lot of nitpicks.

tomereli commented:
gitlab failed due to `RECONF` failure in the rax40 - this sometimes happens, we need to restart the RAX40 and re-run the test - @rmelotte.
```

ERROR 14:17:38:926 <1988345080> ap_wlan_hal_dwpal.cpp[1117] --> Autoconfiguration: "RECONF wlan2.0" command to hostapd has failed!
DEBUG 14:17:38:930 <1988345080> base_wlan_hal_dwpal.cpp[501] --> 
cmd failed: RECONF wlan2.1
DEBUG 14:17:38:931 <1988345080> base_wlan_hal_dwpal.cpp[502] --> 
reply: FAIL

ERROR 14:17:38:932 <1988345080> ap_wlan_hal_dwpal.cpp[1117] --> Autoconfiguration: "RECONF wlan2.1" command to hostapd has failed!
DEBUG 14:17:38:936 <1988345080> base_wlan_hal_dwpal.cpp[501] --> 
cmd failed: RECONF wlan2.2
DEBUG 14:17:38:937 <1988345080> base_wlan_hal_dwpal.cpp[502] --> 
reply: FAIL

ERROR 14:17:38:938 <1988345080> ap_wlan_hal_dwpal.cpp[1117] --> Autoconfiguration: "RECONF wlan2.2" command to hostapd has failed!
DEBUG 14:17:38:942 <1988345080> base_wlan_hal_dwpal.cpp[501] --> 
cmd failed: RECONF wlan2.3
DEBUG 14:17:38:942 <1988345080> base_wlan_hal_dwpal.cpp[502] --> 
reply: FAIL
```

Commit:
bwl: add fronthaul and backhaul to VAPElement

Preparative commit.
The VAP type can be fronthaul, backhaul or hybrid.
In order to read it in refresh_radio_info, add two more boolean values
to the VAPElement struct.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: update VAP type in refresh_vap_info

Read the VAP type in refresh_vap_info using GET_MESH_MODE command in
dwpal variant. Add a TODO with a link to the task in which this will be
added to the nl80211 BWL variant.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ap_manager: remove backhaul_vaps_list

Now that the backhaul vaps are read as part of refresh_vap_info in bwl,
this list is no longer necessary - the previous assumption was that the
backhaul vaps configuration is stored in the platform DB, which is
incorrect for EasyMesh since the configuration should be received from
the controller, and each VAP can support both bAP and fAP configuration.

Since previously each bAP could support only 1 IRE, the agent lofic was
to enable a new bAP every time an IRE connects to the current bAP, so
effectively there is always one active bAP beaconing.

This is again not needed, since the driver (which will be integrated
soon enough) can support any number of stations.

Delete backhaul_vaps_list from the AP manager, and any
use of it throughout the code.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: add backhaul wps ssid and password to update_vap_credentials

Preparative commit.
Add backhaul wps ssid and passphrase to the update_vap_credentials BWL
API which will be used by the BWL to configure the first fronthaul VAP
with the backhaul credentials for WPS onboarding.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: dwpal: use intel mesh_mode

Intel hostapd doesn't support upstream hostapd multi_ap configuration
due to a hardware bug which does not allow configuring VLANs.
Instead, Intel provides "mesh_mode" configuration which allows
configuring either fAP or bAP (default is bAP).
In bAP mode, only 4addr stations which include the WFA multiap vendor
element (0x1B) will be accepted.

Move multiap configuration to a static function which first sets
start_disabled=0, then handles each supported bss_type.

In addition, add set_multiap_wps() private function which sets WPS
onboarding according to the easymesh specification if a backhaul and
fronthaul VAPs are configured. If only fronthaul VAP is configured, only
WPS for fronthaul will be supported.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: add start wps request yaml files

Preparative commit.
Add cACTION_BACKHAUL_START_WPS_PBC_REQUEST and
ACTION_BACKHAUL_START_WPS_PBC_REQUEST to backhaul manager yamls which
will be used to trigger WPS PBC on the 2.4G or 5G radios.

Similarly, Add cACTION_APMANAGER_START_WPS_PBC_REQUEST and
ACTION_APMANAGER_START_WPS_PBC_REQUEST yamls.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: update autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: add backhaul start_wps_pbc and datapath

Preparative commit.
Add start_wps_pbc API in the backhaul manager which is called by the
agent ucc listener on startWpsPbc command, and pass the message till it
gets to the ap manager.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: bwl: add wps pbc impl and datapath

Add start_wps_pbc() API to bwl which starts wps on the given interface,
and update the agent to call the newly added API when the request
is received in the ap manager datapath.

Note that Intel hostapd (and therefore BWL DWPAL) supports WPS PBC on
specific BSS only, unlike upstream hostapd which takes no argument.
For this reason, we trigger WPS PBC on the first fronthaul interface.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: configData set bss_type to uint8_t type

Preparative commit.
The bss_type cannot be an enum, since the enum
(eWscVendorExtSubelementBssType) holds the bitmask values for the
bss_type element - for example, a BSS can be configured to hybrid mode,
e.g. setting BACKHAUL_BSS and FRONTHAUL_BSS bits.
Since enum can contain a single value, change the bss_type parameter to
uint8_t.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bcl: sBssInfoConf replace bss_type with boolean values

As pointed out in
https://github.com/prplfoundation/prplMesh/pull/1168#discussion_r411267323,
it is better to split up the fronthaul and backhaul as separate booleans
in the sBssInfoConf structur. This saves the confusion about the
teardown and bSTA bits, and we don't have to "parse" the bss_type in
every bwl implementation, so do this now.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
test_flows.py: add wps test to autoconfig renew

We can't realy test wireless backhaul in simulation, but we can at least
test the wps CAPI command.

- Add start_wps_registration CAPI command to capi.py
- Add convinience function for it in environment.py
- Trigger WPS on the 2.4G radio as the last step in auto_config_renew

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: add MAP-4.2.3 (wps onboarding) eth test variants

Now that we officially pass the WPS onboarding using ethernet backhaul,
add the tests to the nightly expected passing agent tests.

MAP-4.2.3_ETH_FH24G:netgear-rax40
MAP-4.2.3_ETH_FH5GL:netgear-rax40
MAP-4.2.3_ETH_FH5GH:netgear-rax40

Make sure that normal onboarding still works:

MAP-4.2.1:netgear-rax40

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -825,6 +825,56 @@ bool base_wlan_hal_dwpal::refresh_radio_info()
+    std::string cmd = "GET_MESH_MODE " + ifname;
+
+    fronthaul = false;
+    backhaul  = false;
```
I think there's no need to initialize these variables to `false`, since if this method returns `false` the values should be ignored anyway (nitpick)

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -880,10 +930,16 @@ bool base_wlan_hal_dwpal::refresh_vap_info(int vap_id)
     LOG(DEBUG) << "Detected VAP ID (" << vap_id << ") - MAC: " << vapElement.mac
-               << ", SSID: " << vapElement.ssid;
+               << ", SSID: " << vapElement.ssid
+               << ", fronthaul: " << beerocks::string_utils::bool_str(vapElement.fronthaul)
+               << ", backhaul: " << beerocks::string_utils::bool_str(vapElement.backhaul);
```
Why not use [`std::boolalpha`](https://en.cppreference.com/w/cpp/io/manip/boolalpha) instead?
```suggestion
               << ", fronthaul: " << std::boolalpha << vapElement.fronthaul
               << ", backhaul: " << vapElement.backhaul;
```

vitalybu commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -877,7 +877,7 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
     case beerocks_message::ACTION_APMANAGER_START_WPS_PBC_REQUEST: {
         LOG(DEBUG) << "Got ACTION_APMANAGER_START_WPS_PBC_REQUEST";
-        //TODO
+        ap_wlan_hal->start_wps_pbc();
```
It can fail, so better check the return value

Commit:
bwl: dwpal: Enable VAP on configuration success

It may be that the VAP was previousely disabled, then it is enabled, in
which case the start_disabled will not be removed.
Fix this by explicitely removing the start_disabled flag for
reconfigured VAPs.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
transport: remove packet from de-deuplication map on forward error

When a packet is to be forwarded by the ieee1905 transport, it first
goes through the de-duplication step, in which it is dropped if it
matches a previous packet which was handled by the transport.

Only after that, the message sent to the forwarder, which may fail,
resulting with the same message sent from the source, which will now be
dropped indefinitely since it is found in the de-duplication map.

In order to fix this wrong behaviour -
1. update forward_packet to return false on error and when the packet is
dropped
2. Remove packets which failed forwarding from the de-duplication map so
they won't be dropped if and when they are resent by the source.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
Merging manually since the CI is stuck. After analysis it appears that when 4.2.3 tests fail, any message from the controller is not received by the DUT at all. Will open a separate issue on this.
